File: sol/1277/1277.cpp
struct Solution {
    int countSquares(vector<vector<int>>& matrix) {
        int m = matrix.size(), n = matrix[0].size(), res = 0, buff, newBuff;
        vector<int> memo(matrix[0]);
        for (int j = 0; j < n; ++j){
            res += memo[j];
        }
        for (int i = 1; i < m; ++i){
            buff = memo[0];
            memo[0] = matrix[i][0];
            res += memo[0];
            for (int j = 1; j < n; ++j){
                newBuff = 0;
                if (matrix[i][j] == 1){
                    newBuff = min(buff, min(memo[j - 1], memo[j])) + 1;
                }
                buff = memo[j];
                memo[j] = newBuff;
                res += memo[j];
            }
        }
        return res;
    }
};

File: sol/1277/1277.py
class Solution:
    def countSquares(self, matrix: List[List[int]]) -> int:
        m = len(matrix)
        n = len(matrix[0])
        res = sum(matrix[0])
        for i in range(1, m):
            for j in range(1, n):
                if matrix[i][j] == 1:
                    matrix[i][j] = min(matrix[i - 1][j - 1], matrix[i - 1][j], matrix[i][j - 1]) + 1
            res += sum(matrix[i])
        return res
